src/Symfony/Component/Validator/Constraints/NotPwnedValidator.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Exception\RuntimeException;
+use Symfony\Component\Validator\Exception\UnexpectedTypeException;
+
+/**
+ * Checks if a password has been leaked in a data breach using haveibeenpwned.com's API.
+ * Use a k-anonymity model to protect the password being searched for.
+ *
+ * @see https://haveibeenpwned.com/API/v2#SearchingPwnedPasswordsByRange
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+class NotPwnedValidator extends ConstraintValidator
+{
+    private const RANGE_API = 'https://api.pwnedpasswords.com/range/%s';
+    private $httpClient;
+
+    /**
+     * @param callable|null $httpClient function(string $url): string (throw an exception in case of error)
+     */
+    public function __construct(?callable $httpClient = null)
+    {
+        $this->httpClient = $httpClient;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function validate($value, Constraint $constraint)
+    {
+        if (!$constraint instanceof NotPwned) {
+            throw new UnexpectedTypeException($constraint, NotPwned::class);
+        }
+
+        if (null === $value || '' === $value) {
+            return;
+        }
+
+        if (\is_array($value) || (\is_object($value) && !method_exists($value, '__toString'))) {
```

Comment:
should look like https://github.com/symfony/symfony/blob/5124432b649a0a11c04eeac424d64bbc1aa5c44f/src/Symfony/Component/Validator/Constraints/EmailValidator.php#L77 no?